esp-hal/src/sdio/pins.rs

```diff
@@ -0,0 +1,181 @@
+use super::{Error, Mode};
+use crate::gpio::{Flex, Level, Output, OutputConfig, OutputPin, Pin};
+
+/// Represents the GPIO pins used for SDIO communication.
+///
+/// The pins are configurable for the SPI + SD (1-bit, 4-bit) operation modes.
+///
+/// The CLK/SCLK pin is configured as an output, all other pins are input/output
+/// `Flex` pins.
+#[derive(Debug)]
+pub struct Pins<'d> {
+    mode: Mode,
+    pub(crate) clk_sclk: Output<'d>,
+    pub(crate) cmd_mosi: Flex<'d>,
+    pub(crate) dat0_miso: Flex<'d>,
+    pub(crate) dat1_irq: Flex<'d>,
+    pub(crate) dat2: Flex<'d>,
+    pub(crate) dat3_cs: Flex<'d>,
+}
+
+impl<'d> Pins<'d> {
+    /// Creates a new [Pins] from the provided GPIO pins.
+    ///
+    /// ## Example
+    ///
+    /// ```rust, no_run
+    #[doc = crate::before_snippet!()]
+    /// use esp_hal::sdio::{Mode, Pins};
+    ///
+    /// let _pins = Pins::new(
+    ///     Mode::Sd4bit,
+    ///     peripherals.GPIO19, // CLK/SCLK
+    ///     peripherals.GPIO18, // CMD/MOSI
+    ///     peripherals.GPIO20, // DAT0/MISO
+    ///     peripherals.GPIO21, // DAT1/IRQ
+    ///     peripherals.GPIO22, // DAT2
+    ///     peripherals.GPIO23, // DAT3/#CS
+    /// );
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new(
+        mode: Mode,
+        clk_sclk: impl OutputPin + 'd,
+        cmd_mosi: impl Pin + 'd,
+        dat0_miso: impl Pin + 'd,
+        dat1_irq: impl Pin + 'd,
+        dat2: impl Pin + 'd,
+        dat3_cs: impl Pin + 'd,
+    ) -> Self {
+        Self {
+            mode,
+            clk_sclk: Output::new(clk_sclk, Level::Low, OutputConfig::default()),
+            cmd_mosi: Flex::new(cmd_mosi),
+            dat0_miso: Flex::new(dat0_miso),
+            dat1_irq: Flex::new(dat1_irq),
+            dat2: Flex::new(dat2),
+            dat3_cs: Flex::new(dat3_cs),
```

Comment:
Why do you need to create pin drivers?

Reply:
Because, at least in `SDIO` mode, some of these pins act as both input and output. Is there a better way to express that using the `PeripheralInput/Output` types?

Reply:
The API is meant to take `impl PeripheralInput` and `impl PeripheralOutput`. Internally, drivers are supposed to use gpio::interconnect types, namely `OutputSignal` and `InputSignal`. You can take a look at, say, I2C for an example.

Reply:
Just noting that OutputSignal is a slight misnomer as it is an input/output signal in truth.

Reply:
After reviewing the available `InputSIgnal/OutputSignal` definitions in `esp-metadata`, there don't appear to be signal definitions for `SDIO_CLK`, `SDIO_CMD`, or `SDIO_DATA[0-3]` in the `esp32` or `esp32c6` TOML files. The ESP32-C6 TRM also doesn't have those signals defined. How should I proceed?

Reply:
let us know if you can find this in esp-idf, otherwise we'll ask around internally

Reply:
There aren't signals for the SDIO, it's IO_MUX only, like the USB_OTG signals. Have a look at the USB driver for inspiration.

Reply:
I ended up finding the signals in the ESP32 config in `esp-metadata`. The config just leaves out the signal number, so that is what I did for the ESP32-C6 config. `SD_CMD` was also missing from the input signals for ESP32, so I added it since the command line is full duplex I/O.

Reply:
OH